Upgrade cert-manager to v0.13.1

munnerz opened the pull request:
This pull request upgrades the cert-manager installation to v0.13.1.

I have split it into three parts to indicate the steps that should be applied to the cluster in order to upgrade.

It can be summed up as uninstalling and reinstalling cert-manager. This is because we have change the API group used for our CRD resource (to avoid the `k8s.io` suffix) and there isn't a way for CRDs to migrate between API groups.

I'd recommend *not* deleting the `cert-manager` namespace as it will contain the Let's Encrypt account private key(s) used by ClusterIssuer resources (the `--cluster-resource-namespace` flag defaults to `cert-manager`, and cluster scoped resources store there 'referenced objects' in this namespace as they are cluster scoped themselves 😄).

It should be safe to delete the CRD resources (which will cause all existing configured Certificate/ClusterIssuer/Order/Challenge resources to be deleted) as we have a copy of all of these in this repository (and cert-manager can rebuild the state and 'adopt' the existing issued certificates just fine).

ps. this is a big blob of YAML (6k lines!) due to the OpenAPI validation schema... 🤷‍♂ 

/cc @thockin

munnerz commented:
/area cert-manager

munnerz commented:
/hold

until after #582

munnerz commented:
The actual diff for all this stuff is a bit wild due to us re-organising the Helm chart that is the source for these manifests in the upstream repo (we moved away from using nested-subcharts and now store everything in one chart, but that's 'reordered' how things are deployed as a result). Future diffs should be less unwieldy 🤞

bartsmykla commented:
@munnerz it's hard ro do proper code review for the whole 6k+ lines, so briefly it looks ok. I think we need to be very carefull applying that kind of changes, so I trust you checked this multiple times. :-D

PS. Can you do a rebase?

munnerz commented:
I've updated this PR to:

1) rebase it (and made sure to copy across the change in #582: https://github.com/kubernetes/k8s.io/pull/582/files#diff-57025a035f611c051c04edd8f31d7b58R2188)
2) split into two commits where the first attempts to keep the ordering of resources the same for easier review
3) this is a no-op change, simply reordering all the YAML to be in the same order as the 'upstream' `cert-manager.yaml` file, which should prevent awkward diffs in future (so long as the cert-manager project doesn't do something stupid 🙃)

The diff here is still pretty large, and this can be accounted to two things:

1) cert-manager switched to use `controller-tools` schemagen to generate our OpenAPI schema, which resulted in a bit of 'reordering' of fields within the validation schema
2) upon switching, we found that we'd actually been *missing* some parts of the validation schema due to a bug in our previous generator. This is now fixed, and resulted in ~5k LoC being added to our validation schemas 😅 (we now also have `spec.preserveUnknownFields` set to false since we've validated through automated tests that our OpenAPI schema is now _complete_ and structural 🎉)

In terms of applying this change, because it involves the change of API group of the CRDs, it will involve a complete uninstall and then reinstall. This will not affect any deployed services as 'ownerReferences' are *NOT* set on the Secret resources that cert-manager generates/manages.

I'm happy to apply this change, but ahead of working on AAA figure it'd be good to make sure everyone is on the same page.

bartsmykla commented:
@munnerz thank you for your work on it and clarification! It's much appreciated!

@dims @thockin please look at this and also could you write HOW and mostly WHO should proceed with the @munnerz suggestion:
> In terms of applying this change, because it involves the change of API group of the CRDs, it will involve a complete uninstall and then reinstall. This will not affect any deployed services as 'ownerReferences' are NOT set on the Secret resources that cert-manager generates/manages.

/lgtm
/unhold
/assign @dims @thockin

bartsmykla commented:
/hold

munnerz commented:
> @dims @thockin please look at this and also could you write HOW and mostly WHO should proceed with the @munnerz suggestion:

To clarify here, I'm happy to pick this up and do it 😄 I already have permissions in the aaa cluster to do it, and can spare the time. Let me know what works best for everyone.

bartsmykla commented:
@munnerz I think it's a great idea, let's wait for some comments from @dims or/and @thockin but I'm +1 with the idea of doing it by yourself (as an expert ofc!)

dims commented:
@munnerz go for it!

/approve
/lgtm

bartsmykla commented:
/unhold
